File: backend/src/services/ai-ops/topology/diffEngine.ts
/**
 * 差分计算引擎
 *
 * Property 9: 增量差分幂等性
 * Requirements: 4.1-4.5
 */

import { TopologyGraph, TopologyDiff, TopologyNode, TopologyEdge } from './types';

/**
 * 比较两个 TopologyGraph，生成 TopologyDiff
 */
export function computeDiff(oldGraph: TopologyGraph, newGraph: TopologyGraph): TopologyDiff {
  const now = Date.now();
  const nodesAdded: TopologyNode[] = [];
  const nodesRemoved: TopologyNode[] = [];
  const edgesAdded: TopologyEdge[] = [];
  const edgesRemoved: TopologyEdge[] = [];
  const edgesUpdated: { edgeId: string; changes: Record<string, { old: unknown; new: unknown }> }[] = [];
  const nodesUpdated: { nodeId: string; changes: Record<string, { old: unknown; new: unknown }> }[] = [];

  // 新增节点：在 newGraph 中但不在 oldGraph 中
  for (const [id, node] of newGraph.nodes) {
    if (!oldGraph.nodes.has(id)) {
      nodesAdded.push(node);
    }
  }

  // 移除节点：在 oldGraph 中但不在 newGraph 中
  for (const [id, node] of oldGraph.nodes) {
    if (!newGraph.nodes.has(id)) {
      nodesRemoved.push(node);
    }
  }

  // 新增边：在 newGraph 中但不在 oldGraph 中
  for (const [id, edge] of newGraph.edges) {
    if (!oldGraph.edges.has(id)) {
      edgesAdded.push(edge);
    }
  }

  // 移除边：在 oldGraph 中但不在 newGraph 中
  for (const [id, edge] of oldGraph.edges) {
    if (!newGraph.edges.has(id)) {
      edgesRemoved.push(edge);
    }
  }

  // 属性变更节点：两个图中都存在但属性不同
  for (const [id, newNode] of newGraph.nodes) {
    const oldNode = oldGraph.nodes.get(id);
    if (!oldNode) continue;

    const changes: Record<string, { old: unknown; new: unknown }> = {};
    if (oldNode.state !== newNode.state) {
      changes.state = { old: oldNode.state, new: newNode.state };
    }
    if (oldNode.hostname !== newNode.hostname) {
      changes.hostname = { old: oldNode.hostname, new: newNode.hostname };
    }
    if (JSON.stringify(oldNode.ipAddresses?.sort() || []) !== JSON.stringify(newNode.ipAddresses?.sort() || [])) {
      changes.ipAddresses = { old: oldNode.ipAddresses, new: newNode.ipAddresses };
    }
    if (oldNode.stabilityTier !== newNode.stabilityTier) {
      changes.stabilityTier = { old: oldNode.stabilityTier, new: newNode.stabilityTier };
    }

    if (Object.keys(changes).length > 0) {
      nodesUpdated.push({ nodeId: id, changes });
    }
  }

  // 属性变更边：两个图中都存在但属性不同
  for (const [id, newEdge] of newGraph.edges) {
    const oldEdge = oldGraph.edges.get(id);
    if (!oldEdge) continue;

    const changes: Record<string, { old: unknown; new: unknown }> = {};
    if (oldEdge.confidence !== newEdge.confidence) {
      changes.confidence = { old: oldEdge.confidence, new: newEdge.confidence };
    }
    if (oldEdge.localInterfaceRunning !== newEdge.localInterfaceRunning) {
      changes.localInterfaceRunning = { old: oldEdge.localInterfaceRunning, new: newEdge.localInterfaceRunning };
    }
    if (oldEdge.remoteInterfaceRunning !== newEdge.remoteInterfaceRunning) {
      changes.remoteInterfaceRunning = { old: oldEdge.remoteInterfaceRunning, new: newEdge.remoteInterfaceRunning };
    }
    if (oldEdge.state !== newEdge.state) {
      changes.state = { old: oldEdge.state, new: newEdge.state };
    }
    if (JSON.stringify(oldEdge.sources.sort()) !== JSON.stringify(newEdge.sources.sort())) {
      changes.sources = { old: oldEdge.sources, new: newEdge.sources };
    }

    if (Object.keys(changes).length > 0) {
      edgesUpdated.push({ edgeId: id, changes });
    }
  }

  return {
    id: `diff-${now}`,
    timestamp: now,
    nodesAdded, nodesRemoved, nodesUpdated,
    edgesAdded, edgesRemoved, edgesUpdated,
  };
}

/**
 * 将 TopologyDiff 增量应用到 TopologyGraph
 */
export function applyDiff(graph: TopologyGraph, diff: TopologyDiff): TopologyGraph {
  // 深拷贝 Map
  const nodes = new Map(graph.nodes);
  const edges = new Map(graph.edges);

  // 添加新节点
  for (const node of diff.nodesAdded) {
    nodes.set(node.id, node);
  }

  // 移除节点
  for (const node of diff.nodesRemoved) {
    nodes.delete(node.id);
  }

  // 应用节点属性变更
  for (const update of diff.nodesUpdated || []) {
    const node = nodes.get(update.nodeId);
    if (!node) continue;
    const updated = { ...node };
    for (const [key, change] of Object.entries(update.changes)) {
      (updated as Record<string, unknown>)[key] = change.new;
    }
    nodes.set(update.nodeId, updated);
  }

  // 添加新边
  for (const edge of diff.edgesAdded) {
    edges.set(edge.id, edge);
  }

  // 移除边
  for (const edge of diff.edgesRemoved) {
    edges.delete(edge.id);
  }

  // 应用边属性变更
  for (const update of diff.edgesUpdated) {
    const edge = edges.get(update.edgeId);
    if (!edge) continue;
    const updated = { ...edge };
    for (const [key, change] of Object.entries(update.changes)) {
      (updated as Record<string, unknown>)[key] = change.new;
    }
    edges.set(update.edgeId, updated);
  }

  return {
    nodes, edges,
    version: graph.version + 1,
    lastUpdatedAt: diff.timestamp,
  };
}

/** 检查 diff 是否为空 */
export function isDiffEmpty(diff: TopologyDiff): boolean {
  return diff.nodesAdded.length === 0
    && diff.nodesRemoved.length === 0
    && (diff.nodesUpdated?.length || 0) === 0
    && diff.edgesAdded.length === 0
    && diff.edgesRemoved.length === 0
    && diff.edgesUpdated.length === 0;
}

File: opsevo-python/src/opsevo/services/topology/diff_engine.py
"""
拓扑差分引擎 — 比较两个 TopologyGraph 生成 TopologyDiff

Requirements: 16.2
"""

from __future__ import annotations

import time
import uuid
from typing import Any

from opsevo.services.topology.types import (
    TopologyGraph, TopologyDiff, TopologyNode, TopologyEdge,
)


def compute_diff(old_graph: TopologyGraph, new_graph: TopologyGraph) -> TopologyDiff:
    """比较新旧拓扑图，生成差分。"""
    now = time.time()
    nodes_added: list[TopologyNode] = []
    nodes_removed: list[TopologyNode] = []
    nodes_updated: list[dict[str, Any]] = []
    edges_added: list[TopologyEdge] = []
    edges_removed: list[TopologyEdge] = []
    edges_updated: list[dict[str, Any]] = []

    # Nodes
    for nid, node in new_graph.nodes.items():
        if nid not in old_graph.nodes:
            nodes_added.append(node)
        else:
            changes = _diff_node(old_graph.nodes[nid], node)
            if changes:
                nodes_updated.append({"node_id": nid, "changes": changes})
    for nid, node in old_graph.nodes.items():
        if nid not in new_graph.nodes:
            nodes_removed.append(node)

    # Edges
    for eid, edge in new_graph.edges.items():
        if eid not in old_graph.edges:
            edges_added.append(edge)
        else:
            changes = _diff_edge(old_graph.edges[eid], edge)
            if changes:
                edges_updated.append({"edge_id": eid, "changes": changes})
    for eid, edge in old_graph.edges.items():
        if eid not in new_graph.edges:
            edges_removed.append(edge)

    return TopologyDiff(
        id=f"diff-{uuid.uuid4().hex[:8]}",
        timestamp=now,
        nodes_added=nodes_added, nodes_removed=nodes_removed,
        edges_added=edges_added, edges_removed=edges_removed,
        edges_updated=edges_updated, nodes_updated=nodes_updated,
    )


def _diff_node(old: TopologyNode, new: TopologyNode) -> dict[str, dict[str, Any]]:
    changes: dict[str, dict[str, Any]] = {}
    for attr in ("state", "confirm_count", "miss_count", "ip_addresses", "sources"):
        ov = getattr(old, attr)
        nv = getattr(new, attr)
        if ov != nv:
            changes[attr] = {"old": ov, "new": nv}
    return changes


def _diff_edge(old: TopologyEdge, new: TopologyEdge) -> dict[str, dict[str, Any]]:
    changes: dict[str, dict[str, Any]] = {}
    for attr in ("state", "confidence", "confirm_count", "miss_count", "sources",
                 "local_interface_running", "remote_interface_running"):
        ov = getattr(old, attr)
        nv = getattr(new, attr)
        if ov != nv:
            changes[attr] = {"old": ov, "new": nv}
    return changes
